File: sdks/python/hatchet_sdk/features/workers.py
import asyncio

from hatchet_sdk.clients.rest.api.worker_api import WorkerApi
from hatchet_sdk.clients.rest.api_client import ApiClient
from hatchet_sdk.clients.rest.models.update_worker_request import UpdateWorkerRequest
from hatchet_sdk.clients.rest.models.worker import Worker
from hatchet_sdk.clients.rest.models.worker_list import WorkerList
from hatchet_sdk.clients.rest.tenacity_utils import tenacity_retry
from hatchet_sdk.clients.v1.api_client import BaseRestClient


class WorkersClient(BaseRestClient):
    """
    The workers client is a client for managing workers programmatically within Hatchet.
    """

    def _wa(self, client: ApiClient) -> WorkerApi:
        return WorkerApi(client)

    def get(self, worker_id: str) -> Worker:
        """
        Get a worker by its ID.

        :param worker_id: The ID of the worker to retrieve.
        :return: The worker.
        """
        with self.client() as client:
            worker_get = tenacity_retry(
                self._wa(client).worker_get, self.client_config.tenacity
            )
            return worker_get(worker_id)

    async def aio_get(self, worker_id: str) -> Worker:
        """
        Get a worker by its ID.

        :param worker_id: The ID of the worker to retrieve.
        :return: The worker.
        """
        return await asyncio.to_thread(self.get, worker_id)

    def list(
        self,
    ) -> WorkerList:
        """
        List all workers in the tenant determined by the client config.

        :return: A list of workers.
        """
        with self.client() as client:
            worker_list = tenacity_retry(
                self._wa(client).worker_list, self.client_config.tenacity
            )
            return worker_list(
                tenant=self.client_config.tenant_id,
            )

    async def aio_list(
        self,
    ) -> WorkerList:
        """
        List all workers in the tenant determined by the client config.

        :return: A list of workers.
        """
        return await asyncio.to_thread(self.list)

    def update(self, worker_id: str, opts: UpdateWorkerRequest) -> Worker:
        """
        Update a worker by its ID.

        :param worker_id: The ID of the worker to update.
        :param opts: The update options.
        :return: The updated worker.
        """
        with self.client() as client:
            return self._wa(client).worker_update(
                worker=worker_id,
                update_worker_request=opts,
            )

    async def aio_update(self, worker_id: str, opts: UpdateWorkerRequest) -> Worker:
        """
        Update a worker by its ID.

        :param worker_id: The ID of the worker to update.
        :param opts: The update options.
        :return: The updated worker.
        """
        return await asyncio.to_thread(self.update, worker_id, opts)

File: sdks/ruby/src/lib/hatchet/features/workers.rb
# frozen_string_literal: true

module Hatchet
  module Features
    # Workers client for managing workers programmatically within Hatchet
    #
    # This class provides a high-level interface for retrieving, listing, and
    # updating workers in the Hatchet system.
    #
    # @example Getting a worker
    #   worker = workers_client.get("worker-id")
    #
    # @example Listing all workers
    #   workers = workers_client.list
    #
    # @since 0.1.0
    class Workers
      # Initializes a new Workers client instance
      #
      # @param rest_client [Object] The configured REST client for API communication
      # @param config [Hatchet::Config] The Hatchet configuration containing tenant_id and other settings
      # @return [void]
      # @since 0.1.0
      def initialize(rest_client, config)
        @rest_client = rest_client
        @config = config
        @worker_api = HatchetSdkRest::WorkerApi.new(rest_client)
      end

      # Get a worker by its ID
      #
      # @param worker_id [String] The ID of the worker to retrieve
      # @return [Object] The worker details
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   worker = workers_client.get("worker-123")
      def get(worker_id)
        @worker_api.worker_get(worker_id)
      end

      # List all workers in the tenant
      #
      # @return [Object] A list of workers
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   workers = workers_client.list
      def list
        @worker_api.worker_list(@config.tenant_id)
      end

      # Update a worker by its ID
      #
      # @param worker_id [String] The ID of the worker to update
      # @param opts [Hash] The update options
      # @return [Object] The updated worker
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   updated = workers_client.update("worker-123", { is_paused: true })
      def update(worker_id, opts)
        update_request = HatchetSdkRest::UpdateWorkerRequest.new(opts)
        @worker_api.worker_update(worker_id, update_request)
      end
    end
  end
end
